### ISSUE TITLE
`satisfies` shouldn't complain about unknown properties

### ISSUE THREAD

**Issue by mmkal** [microsoft/TypeScript#52999]
# Suggestion

## 🔍 Search Terms

satisfies + (known properties, unknown properties, excess)

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

When using `satisfies`, the value should be allowed to have extra properties.

## 📃 Motivating Example

`Satisfy` is defined this way when I Google it:
>meet the expectations, needs, or desires of (someone).

Right now, the `satisfies` operator does more than that. It's more like `satisfies and goes no further than`. Example:

```ts
interface User {
  name: string
}

const john = {
  name: 'John',
  email: 'john@gmail.com',
} satisfies User
```

In the above, the value of `john` clearly does satisfy the interface `User`, but the compiler complains:

```
Type '{ name: string; email: string; }' does not satisfy the expected type 'User'.
  Object literal may only specify known properties, and 'email' does not exist in type 'User'.
```

The first line of this is straight-up untrue! And it'd be really useful to be able to say "I have an object which I know is assignable to `User`, but it also has some specific extra stuff which I still care about.

## 💻 Use Cases

This first arose writing a very large, nested, complex config for a CMS. Right now it's a huge `as const` expression, which we have to carefully align with a type `Config` - but the type only defines _some_ of the properties a config can have at compile time. The value is later passed to a function which expects a valid `Config` so we see invalid configuration errors in the wrong place. We have tried to use `satisfies` but hit the excess properties error for those properties not defined on the `Config` type. It's nested so we can't do `satisfies Config & Record<string, unknown>` except for top-level properties.

**Comment by martinwepner** [comment]
For me and probably many others its important that `satisfies` works as it does.
However have you tried the following? For your use-case I always check like this:

```typescript
interface User {
  name: string
}

const john = {
  name: 'John',
  email: 'john@gmail.com',
}

// compile time check if john is assignable to User
;(x: typeof john): User  => x


// or in your case:
type Config = {
  key1: string
} // & { <A_COMPLEX_TYPE> }
const config = {
  key1: "config1",
  // keys with values of A_COMPLEX_TYPE
  keyNotInConfig: "foo"
}

// compile time check if config is assignable to Config
;(x: typeof config): Config  => x
```

[Playground link](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgKoGdrIN4ChnIhwC2EAXMumFKAOa4C+uuCA9iFcgFasAWIyALw58hEuWQByAFJ8QkgDSiIxOMAA2FST34ABWqo0A6NsUWNmAekvJTABw0owwUrd4QEAa2TAY3OT7oyHDo6MC0RABG6k6saJhQuADcABQAHhRgAJ52EKx+OiAAlBQYWEIAfMhpzLjWyKxQPgJZrACuTQgh5LjZucgAwuww4UIiBJ4QWQCMFFQ0IPQMyPUAZDjIADwAggD6AwDyALIACgAyAKIAGrsAKgCaJxdVTGwcYLbDo8J4E1OzyAARG8RrRpoClAR6pMskEAO7AMC8ZAANzg6jaECC+WQe0Op0uNweT1EMIAcqwwABJEBDECgiiAmCsViAix1Gz2RzIZyuBDuLw+Pwg0bAIIhMIRODRWKDL70VIZHk5PLC+UlOX00aVapAA)


PS: If you want to check for equivalence  you can do this:
```typescript
interface User {
  name: string
}

const john = {
  name: 'John',
  email: 'john@gmail.com',
}

// compile time check if john is equivalent to User
;(x: typeof john): User  => x
;(x: User): typeof john  => x
```

[Playground link](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgKoGdrIN4ChnIhwC2EAXMumFKAOa4C+uuCA9iFcgFasAWIyALw58hEuWQByAFJ8QkgDSiIxOMAA2FST34ABWqo0A6NsUWNmAekvJTABw0owwUrd4QEAa2TAY3OT7oyBAAjgCuwABucOoQ4MhgrGiYULgA3AAUAB4UYACedhCsfjogAJQUGFhCAHzIWenZlSkVCQVFJQG19cxAA)

**Comment by MartinJohns** [comment]
This is intentional. See #47920.

**Comment by mmkal** [comment]
@MartinJohns thank you, I missed that this was intentional behaviour first time I read that.

Re those workarounds:

`satisfies T & Record<string, unknown>`:

Does not work for nested types. The `Config` type I mentioned above looks something like:

```ts
interface Config {
  collections: Array<{name: string; folder: string; fields: Field[]}>
  backend: {repo: string; branch: string}
}

type Field = {name: string; translate?: boolean} & (
  | {widget: 'string'}
  | {widget: 'markdown'}
  | {widget: 'number'}
  | {widget: 'object'; fields: Field[]}
  | {widget: 'list'; field: Field}
  | {widget: 'list'; fields: Field[]}
  | {widget: 'select'; options: string[]}
  | {widget: 'code'}
  // | ...many more options
)
```

Where there are lots of complex field types in the `Field` union - and note that they can go arbitrarily nested via the `object` and `list` field types. So using `& Record<string, unknown>` isn't practical since most of those field types also allow excess properties.

---

`;(x: typeof john): User  => x`:

This works, in that it gives errors for invalidly-defined values, but the reason I made this issue is I want to be able to take advantage of `satisfies` - I want developers to get autocomplete when typing fields, etc. This workaround loses all of that and puts the error on a random, separate, deleteable-looking runtime function. I thought that kind of no-op identity function that is useful at compile time only was what `satisfies` was partly seeking to eliminate.

---

One workaround that really does work and doesn't have any noticeable downsides:

```
interface User {
  name: string
}

const john = {
  name: 'John',
  // @ts-expect-error
  email: 'john@gmail.com',
  birthDate: '2000-01-01',
} satisfies User
```

Since luckily, the compiler only looks for the _first_ excess property. But it seems pretty strange that typescript would push users into doing something like this.

---

Having said that, I do understand that there was a reason to disallow excess properties by default as explained in #47920. I've retitled this issue to "Make `satisfies` configurable to allow excess properties" - maybe there could be some kind of syntax like `satisfies extends User` to allow for this use case? @RyanCavanaugh curious if there was any internal discussion of allowing excess properties via extra syntax? Or is there a better workaround than `// @ts-expect-error`?

**Comment by RyanCavanaugh** [comment]
`ts-expect-error` should always be your last resort; do literally anything else first.

If you're writing this construct exactly once (or some very small number of times) in your codebase, which it sounds like you are, I think it's just way better to write the slightly awkward variant:
```ts
const config = { /* huge thing */ } as const;
config satisfies Config;
```
You won't get error spans inside the object literal, but the error message will give you an unambiguous key path to the offending property if the literal is `as const`.

>  if there was any internal discussion of allowing excess properties via extra syntax

Of course, but generally we don't want to "spend" syntax unless it's absolutely critical to the language.

**Comment by mmkal** [comment]
>If you're writing this construct exactly once (or some very small number of times) in your codebase, which it sounds like you are, I think it's just way better to write the slightly awkward variant:

Yes there's only one config, but it's updated frequently. So it isn't just error spans, it's getting autocomplete on `Field` properties, etc. Developers have found the config hard to understand and I think intellisense would help a lot.

>we don't want to "spend" syntax unless it's absolutely critical to the language

I don't know if I can make the case that it's absolutely critical to the language. But without a way to do this it does seem like perfectly-innocent users doing legitimate things are being forced to choose between bad DX and writing bad code like ts-expect-error.

**Comment by fatcerberus** [comment]
I admit I'm a bit amused by this issue - it seems that the majority of people reporting issues with excess property checks think these checks are a core part of the type system and expect them to be much more comprehensive than they are (hence #12936), while *you* actually want a way to *relax* them. 😄

**Comment by mmkal** [comment]
Yes. We are large, we contain multitudes.

**Comment by martinwepner** [comment]
Ah nice!
```typescript
const config = { /* huge thing */ } as const;
config satisfies Config;
// is much better than 
;(x: typeof config): Config => x
```

Concerning the autocomplete, what about this:
```typescript
type Config = { /* huge thing */ }
const Config = <TConfig extends Config>(config: TConfig) => config
const config = Config({ notInConfig: 42 })
```

**Comment by RyanCavanaugh** [comment]
@dkrieger consider if you wrote something like this
```ts
interface Dimensions {
  width: number;
  height?: number;
  depth?: number;
}

const p = { width: 20, detph: 20 } satisfies Dimensions;
```
The only thing saving you here is excess property checks (EPC). Not erroring on this line is indefensible IMO - if `satisfies` isn't providing this error, it's not doing its stated job.

There's a directionality problem here - if `satisfies` didn't do EPC, there'd be no way to get an error on this line, and the operator is useless toward that goal. But in a world where `satisfies` does EPC (this one), it's something you can opt out of with `& Record<...`. It's obviously better to design it in a way where both scenarios can be satisfied, so that's what we did.

**Comment by alesmenzel** [comment]
@RyanCavanaugh Since `& Record<...` only works one level and is unusable without creating more hacks like the one mentioned by @ackvf . Could you please consider adding new syntax to allow satisfies to disable EPC, since people will hack it anyway.

There is a simple use case that this would be useful
```
type Config = {foo: number}
const config = { foo: 10, bar: '...' } satisfies Config
```

**Comment by RyanCavanaugh** [comment]
> will likely be flagged by any sensible linting configuration as the statement is unused

Why is it "sensible" if it flags unsuspicious code? Linters should work with the user, not against them. What you're saying is that you *have* a syntax that does what you want, but can't use it for reason unrelated to TS.

**Comment by pqnet** [comment]
This behavior is the reason that all the framework expose a `defineConfig` identity function that handle types rather than just a `Config` type. Because of this, things that could simply imported as type only and disappear in tree shaking have to be imported as runtime values and possibly require a bundler/preprocessor in some cases.

**Comment by RyanCavanaugh** [comment]
> This behavior is the reason that all the framework expose a defineConfig identity function

Can you link some examples?

**Comment by pqnet** [comment]
> > This behavior is the reason that all the framework expose a defineConfig identity function
> 
> Can you link some examples?

https://eslint.org/docs/latest/use/configure/configuration-files for example

**Comment by RyanCavanaugh** [comment]
@nzakas / @bradzacher can you explain the above? I don't understand the interaction - it doesn't seem like the EPC behavior of `satisfies` is relevant in that context

**Comment by bradzacher** [comment]
ESLint "flat configs" are just a config file that export an array of objects. It's true that no function is actually required to author one.

The original intent of typescript-eslint's `config` function (which was the precursor to eslint's `defineConfig`) was indeed to provide an identity function for users.

However this wasn't done for `.ts` usage because at the time eslint config files could only have one of the `.js` extensions. So we provided the identity function as a convenient way to provide types to config files without having to mess around with JSDoc comments (which are easy to get wrong - especially when you are trying to add a type to something you are directly exporting).

However once we had this function released we realised it provided a great point for us to add features to - and so we made some improvements to the DevX by introducing runtime code into the function.

Later ESLint core followed suit and adopted our same strategy because it provided many benefits to the DevX. **Their `defineConfig` function has never been a pure identity function.**

For additional context eslint now also supports `.ts` extensions for config files (but that is a recent addition). So you could use `satisfies` and import the type if you wanted - but you would lose the runtime changes the config functions make.

**Comment by pqnet** [comment]
Function arguments can be configured to be of type `T extends MyType` and thus by wrapping the constant value in a call to `<T extends MyType>enforceType(x: T): T => x` (i.e., `const myVar = enforceType({...})`) you can define a constant that:
- is enforced to be compatible with `MyType`
- preserves the most strict typing of the declared constant
- allows additional properties as long as they don't conflict with the properties declared for `MyType`
- provides type mismatch errors (red squiggles) as close as possible to the actual wrong line

The case for big configuration objects is that they often are complex enough that typing them exactly can be difficult, in particular with extensions and plugins that change the functionality and have to deal with a restricted configuration interface. For these thing having a more "lax" typing sometimes can save a lot of hassle.

**Comment by daveycodez** [comment]
Welp I just added [string]: unknown to my parent type. RIP
